File: ops/loop_guard.py
import time
from contextlib import contextmanager
from typing import Any


class LoopGuardAbort(RuntimeError):
    def __init__(self, name: str, iteration: int, elapsed: float, state: dict[str, Any] | None):
        super().__init__(
            f"[LoopGuard] {name} abort at i={iteration}, t={elapsed:.3f}s, state={state}"
        )


class LoopGuard:
    def __init__(
        self,
        *,
        max_iter: int = 10_000_000,
        timeout_s: float = 10.0,
        sample_every: int = 10_000,
        name: str = "loop",
    ) -> None:
        self.max_iter = max_iter
        self.timeout_s = timeout_s
        self.sample_every = sample_every
        self.name = name
        self._i = 0
        self._start = time.perf_counter()

    def tick(self, state: dict[str, Any] | None = None) -> None:
        self._i += 1
        if self._i % self.sample_every != 0:
            return
        elapsed = time.perf_counter() - self._start
        if self._i >= self.max_iter or elapsed >= self.timeout_s:
            raise LoopGuardAbort(self.name, self._i, elapsed, state)


@contextmanager
def loop_guard_context(**kwargs):
    guard = LoopGuard(**kwargs)
    yield guard

File: frontend/src/lib/loopGuard.ts
type LoopGuardOptions = {
  maxIter?: number;
  timeoutMs?: number;
  sampleEvery?: number;
  name?: string;
};

export class LoopGuard {
  private maxIter: number;
  private timeoutMs: number;
  private sampleEvery: number;
  private name: string;
  private count = 0;
  private start = Date.now();
  private lastFingerprint: string | null = null;
  private repeatedSamples = 0;

  constructor({
    maxIter = 10_000,
    timeoutMs = 10_000,
    sampleEvery = 100,
    name = "loop",
  }: LoopGuardOptions = {}) {
    this.maxIter = maxIter;
    this.timeoutMs = timeoutMs;
    this.sampleEvery = sampleEvery;
    this.name = name;
  }

  tick(state: Record<string, unknown> = {}): void {
    this.count += 1;
    if (this.count % this.sampleEvery !== 0) {
      return;
    }
    const fingerprint = this.fingerprint(state);
    if (fingerprint === this.lastFingerprint) {
      this.repeatedSamples += 1;
      if (this.repeatedSamples === 1 || this.repeatedSamples % 5 === 0) {
        // eslint-disable-next-line no-console
        console.warn(`[LoopGuard][state] ${this.name} repeating`, state);
      }
    } else {
      this.repeatedSamples = 0;
      this.lastFingerprint = fingerprint;
      // eslint-disable-next-line no-console
      console.info(`[LoopGuard][state] ${this.name} sample`, state);
    }

    const elapsed = Date.now() - this.start;
    if (this.count >= this.maxIter || elapsed >= this.timeoutMs) {
      const err = new Error(
        `[LoopGuard] ${this.name} abort at i=${this.count}, t=${elapsed}ms, state=${JSON.stringify(state).slice(0, 512)}`,
      );
      (err as Error & { code?: string }).code = "LOOP_GUARD_ABORT";
      throw err;
    }
  }

  reset(): void {
    this.count = 0;
    this.start = Date.now();
    this.lastFingerprint = null;
    this.repeatedSamples = 0;
  }

  private fingerprint(state: Record<string, unknown>): string {
    try {
      const ordered = Object.keys(state)
        .sort()
        .reduce<Record<string, unknown>>((acc, key) => {
          acc[key] = state[key];
          return acc;
        }, {});
      return JSON.stringify(ordered).slice(0, 512);
    } catch {
      return JSON.stringify(Object.keys(state)).slice(0, 512);
    }
  }
}
